MarkBind/markbind
# Add <tree> component for folder structure visualisations

**What is the purpose of this pull request?**




**Overview of changes:**
Fixes #1660

Provides a convenient way to create lightweight tree-like textual output, suitable for folder structure visualisations.
Syntax:
```pre
<tree>
Edit me to generate
  a
    nice
      tree
        diagram!
        :)
</tree>
<tree>
Edit me to generate
  - a
    - nice
    - tree
        - diagram!
        - :)
</tree>
```
To
```pre
Edit me to generate
└── a
    └── nice
        └── tree
            ├── diagram!
            └── :)
```
More details in UG's [tree section](https://deploy-preview-1807--markbind-master.netlify.app/userguide/components/imagesanddiagrams#tree)

A summary of the work done:
- feature development in `core/src/plugins/default/markbind-plugin-tree.js`
- unit test in `core/test/unit/plugins/default/tree.test.js`
- functional test in `packages/cli/test/functional/test_site/testTree.md`
- UG inclusion under Using Component - Images & Diagrams - Tree
- Migration of the existing folder structure code in UG over to use the new tree syntax, mainly in Reusing Contents
- Update tests

**Anything you'd like to highlight / discuss:**
- Initially thought of making a Vue component, but went with making a default plugin instead (since the feature itself is quite lightweight)
- The currently supported syntax uses spaces + unordered list bullets for indication of indent/level, I am not sure whether I should also support (chop off) ordered list bullets (basically, numbers like 1., 2. etc).
- The support for inline markdown/HTML elements works for most syntax, but not for tooltip/popover
```pre
<tooltip id="tt:trigger_id" content="This tooltip triggered by a trigger"></tooltip>

<tree>
C:/course/
  textbook/
    <tooltip content="hello">index.md</tooltip>
    <trigger for="tt:trigger_id">via trigger</trigger>
</tree>
```
Testing the above code will give:
```html
<div class="tree">C:/course/
├── <span data-mb-component-type="tooltip" tabindex="0" content="hello" class="trigger">  index.md</span>
└── <span tabindex="0" class="trigger"> via trigger</span></div>
```

![image](https://user-images.githubusercontent.com/41845017/157072875-ea81f93d-9a5a-4512-b8cc-b81782ca31ab.png)

Notice the white space added in front of the words "index.md" and "via trigger". While the direct tooltip approach does not work on hover, the trigger approach works. Not too sure if this is a bug with the tooltip/trigger element or is it due to the newly added tree component.

**Testing instructions:**
Tests cases are added in
- `packages/cli/test/functional/test_site/testTree.md`
- `packages/core/test/unit/plugins/default/tree.test.js`

User guide section
- [tree](https://deploy-preview-1807--markbind-master.netlify.app/userguide/components/imagesanddiagrams#tree)
- see some migration of the folder structure code in this [page](https://deploy-preview-1807--markbind-master.netlify.app/userguide/reusingcontents#include-inside-an-included-file)



**Proposed commit message: (wrap lines at 72 characters)**
Add <tree> component for folder structure visualisations

It is tedious to manually create tree-like structure diagrams.

Let's add a <tree> component to conveniently convert
an indented list of items into tree-like textual output.

This is especially useful for file system visualisations.



---

**Checklist:** :ballot_box_with_check:
